rb_link/src/server/getter.rs

```diff
@@ -0,0 +1,190 @@
+    pub fn get(&mut self, id: u32) -> B2RMessage {
+        loop {
+            let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+            if let Some(queue) = b2r_cache.get_mut(&id) {
+                if let Some(b2r_message) = queue.pop_front() {
+                    return b2r_message;
+                }
+            }
+            drop(b2r_cache);
+        }
+    }
+
+    /// Get all message send by the probe with id.
+    pub fn get_id_all(&mut self, id: u32) -> Vec<B2RMessage> {
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        let mut messages: Vec<B2RMessage> = Vec::new();
+
+        if let Some(queue) = b2r_cache.get_mut(&id) {
+            while let Some(b2r_message) = queue.pop_front() {
+                messages.push(b2r_message);
+            }
+        }
+        messages
+    }
+
+    /// Return the earliest message from the probe with id.
+    /// This function will return None if there is no message available for retrieval.
+    pub fn try_get(&mut self, id: u32) -> Option<B2RMessage> {
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        if let Some(queue) = b2r_cache.get_mut(&id) {
+            if let Some(b2r_message) = queue.pop_front() {
+                return Some(b2r_message);
+            }
+        }
+        None
+    }
+}
+
+/// A getter that retrieves messages from bluesim sequentially according to cycles.
+pub struct CycleGetter {
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+}
+
+impl CycleGetter {
+    /// Crate a new getter bind to the given server
+    pub fn new(server: &B2RServer) -> Self {
+        CycleGetter {
+            b2r_cache: server.b2r_cache.clone(),
+        }
+    }
+
+    /// Get all the messages sent by the earliest cycle.
+    /// If there are no messages available, it will return an empty Vec.
+    pub fn get_cycle_message(&mut self) -> Vec<B2RMessage> {
+        let mut min_cycles = u32::MAX;
+        let mut messages: Vec<B2RMessage> = Vec::new();
+        let mut b2r_cache = self.b2r_cache.lock().expect("Fail to lock b2r_cache");
+        for queue in b2r_cache.values() {
+            if let Some(b2r_message) = queue.front() {
+                if b2r_message.cycles < min_cycles {
+                    min_cycles = b2r_message.cycles;
+                }
+            }
+        }
+        for queue in b2r_cache.values_mut() {
+            if let Some(b2r_message) = queue.front() {
+                if b2r_message.cycles == min_cycles {
+                    messages.push(queue.pop_front().expect("front error"));
+                }
+            }
+        }
+        messages
+    }
+}
+
+/// The pipeline state
+/// cycle: the cycle of the state
+/// full_fifos: the ids of the full fifos
+/// empty_fifos: the ids of the empty fifos
+/// fire_fules: the fired rules
+pub struct PipeLineState {
+    pub cycle: u32,
+    pub full_fifos: Vec<u32>,
```

Comment:
maybe you should define a new type called ProbeID as u32?

And also, if it stores IDs, then you should show it. Otherwise, when someone read code somewhere else in the code base, and see a struct called `full_fifos`, since there is a `s`, he can known this field is a Vec or a slice like things. But what is in this slice? He may think they are some struct represent a FIFO object, and won't think they are IDs.